### ISSUE TITLE
Submodel Lazy Loading

### ISSUE THREAD

**Issue by will123195** [SkyPHP/skyphp#10]
We should wait to instantiate submodel objects until someone tries to get it.  This should cut down on a lot of memcache traffic.

Something along the lines of this:

``` php
<?
    public function __get($name) {
        if ( $this->propertyExists($name) ) {
            // if we are getting a submodel object, instantiate it if it has not already been loaded
            $object_type = $this->_objects[$name]; // is it possible the object does not extend model?
            if ( $object_type && !$this->_loaded_objects[$name] ) {
                switch ($object_type) {
                case 'plural': // if this property is an array of many objects
                    // load array of objects
                    $this->_date[$name] = array();
                    break;
                default: // if this property is a single object
                    $property = "{$name}_id"; // what if a different constructor id was specified?
                    $this->_data[$name] = new $name( $this->$property );
                }
            }
            $this->_loaded_objects[$name] = true;
            return $this->_data[$name];
        }
        return null;
    }
```

**Comment by stanistan** [comment]
Really good idea. 

The only thing that would be stored in _objects is what is in the AQL of the model, so we don't have to worry about them not being Model objects.

**Comment by stanistan** [comment]
We get the constructor value from the AQL Array, and the result of LoadDB() as an array of the ids to load if it is a plural object, and if it's a single the value should just be an integer.

Will have to rewrite the model class a bit, as well how the AQL select works for objects.

**Comment by will123195** [comment]
We should use mem($array_of_keys) to get all objects of a plural submodel.

**Comment by will123195** [comment]
resolved in 3.0
